### PR TITLE
Define API for Transformer defences and implement Defensive Distillation as example

### LINKED ISSUE #293: Add API and example for transformation defences
Define API for transformation defences which take a trained classifier as input and return a classifier modified to be more robust.

Implement as an example defensive distillation (paper link: https://arxiv.org/pdf/1511.04508.pdf)

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
Fixes #293 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
This PR contributes to issue #251

File: art/defences/__init__.py [hunk 1]
```diff
@@ -14,9 +18,9 @@
...
-from art.defences.postprocessor.gaussian_noise import GaussianNoise
-from art.defences.postprocessor.high_confidence import HighConfidence
-from art.defences.postprocessor.postprocessor import Postprocessor
-from art.defences.postprocessor.reverse_sigmoid import ReverseSigmoid
-from art.defences.postprocessor.rounded import Rounded
+from art.defences.trainer.adversarial_trainer import AdversarialTrainer
+from art.defences.trainer.adversarial_trainer_madry_pgd import AdversarialTrainerMadryPGD
+from art.defences.trainer.trainer import Trainer
+
+from art.defences.transform.distillation import Distillation
```
Comment (beat-buesser) [inline art/defences/__init__.py]:
```suggestion
from art.defences.transform.defensive_distillation import DefensiveDistillation
```

File: art/defences/transform/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+"""
+Module implementing transform-based defences against adversarial attacks.
+"""
+from art.defences.transform.distillation import Distillation
```
Comment (beat-buesser) [inline art/defences/transform/__init__.py]:
```suggestion
from art.defences.transform.defensive_distillation import DefensiveDistillation
```

File: art/defences/transform/distillation.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+import numpy as np
+
+from art.defences.transform.transformer import Transformer
+from art.utils import is_probability
+
+logger = logging.getLogger(__name__)
+
+
+class Distillation(Transformer):
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
class DefensiveDistillation(Transformer):
```

File: art/defences/transform/distillation.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    """
+    This module implements the distillation defence `Distillation`.
+    Implement the distillation defence approach.
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
    Implement the defensive distillation mechanism.
```

File: art/defences/transform/transformer.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the abstract base class for defences that transform one classifier to a more robust classifier.
```
Comment (beat-buesser) [inline art/defences/transform/transformer.py]:
```suggestion
This module implements the abstract base class for defences that transform a classifier into a more robust classifier.
```

File: art/utils.py [hunk 6]
```diff
@@ -816,3 +816,26 @@ def is_probability(vector):
...
+    x = torch.Tensor(np.array([[1, 2, 3], [4, 5, 6]]))
+    y = torch.Tensor(np.array([[1, 2, 3], [4, 5, 6]]))
+
+    try:
+        loss(x, y)
+        result = True
+    except RuntimeError:
+        result = False
+
+    return result
```
Comment (beat-buesser) [inline art/utils.py]: Is this method still required with the currently implemented solution of listing all loss functions?

File: tests/test_utils.py [hunk 7]
```diff
@@ -366,6 +367,15 @@ def test_is_probability(self):
...
 
+    def test_check_pytorch_loss_param(self):
+        import torch.nn as nn
+
+        loss = nn.CrossEntropyLoss()
+        self.assertFalse(check_pytorch_loss_param(loss))
+
+        loss = nn.PoissonNLLLoss()
+        self.assertTrue(check_pytorch_loss_param(loss))
+
```
Comment (beat-buesser) [inline tests/test_utils.py]: Are these tests still required with the currently implemented solution of listing all loss functions?

File: tests/test_utils.py [hunk 8]
```diff
@@ -28,6 +28,7 @@
 from art.utils import second_most_likely_class, random_targets, get_label_conf, get_labels_np_array, preprocess
 from art.utils import segment_by_class, performance_diff
 from art.utils import is_probability
+from art.utils import check_pytorch_loss_param
```
Comment (beat-buesser) [inline tests/test_utils.py]: Is this import still required with the currently implemented solution of listing all loss functions?

File: art/classifiers/pytorch.py [hunk 9]
```diff
@@ -107,6 +107,12 @@ def __init__(
         # Index of layer at which the class gradients should be calculated
         self._layer_idx_gradients = -1
 
+        # Check the loss function for dimension reducing of labels
```
Comment (beat-buesser) [inline art/classifiers/pytorch.py]:
```suggestion
        # Check if the loss function requires as input index labels instead of one-hot-encoded labels
```

File: art/classifiers/tensorflow.py [hunk 10]
```diff
@@ -178,6 +178,10 @@ def fit(self, x, y, batch_size=128, nb_epochs=10, **kwargs):
         # Apply preprocessing
         x_preprocessed, y_preprocessed = self._apply_preprocessing(x, y, fit=True)
 
+        # Check label shape
+        if len(self._labels_ph.shape) == 1:
```
Comment (beat-buesser) [inline art/classifiers/tensorflow.py]: How about moving this condition to __init__ and set with it attribute `self._reduce_labels` to make it consistent across classifiers?

File: art/classifiers/tensorflow.py [hunk 11]
```diff
@@ -311,6 +315,10 @@ def loss_gradient(self, x, y, **kwargs):
         if not hasattr(self, "_loss_grads") or self._loss_grads is None or self._labels_ph is None:
             raise ValueError("Need the loss function and the labels placeholder to compute the loss gradient.")
 
+        # Check label shape
+        if len(self._labels_ph.shape) == 1:
```
Comment (beat-buesser) [inline art/classifiers/tensorflow.py]: If following comment above, we could replace this with `self._reduce_labels`.

File: art/classifiers/tensorflow.py [hunk 12]
```diff
@@ -766,6 +774,10 @@ def fit(self, x, y, batch_size=128, nb_epochs=10, **kwargs):
         # Apply preprocessing
         x_preprocessed, y_preprocessed = self._apply_preprocessing(x, y, fit=True)
 
+        # Check label shape
+        if isinstance(self._loss_object, tf.keras.losses.SparseCategoricalCrossentropy):
```
Comment (beat-buesser) [inline art/classifiers/tensorflow.py]: If following comment above, we could replace this with `self._reduce_labels`.

File: art/defences/transform/distillation.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the distillation defence `Distillation`.
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
This module implements the transforming defence mechanism of defensive distillation.
```

File: art/defences/transform/distillation.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    Implement the distillation defence approach.
+
+    | Paper link: https://arxiv.org/abs/1511.04508
+    """
+
+    params = ["batch_size", "nb_epochs"]
+
+    def __init__(self, classifier, batch_size=128, nb_epochs=10):
+        """
+        Create an instance of the distillation defence.
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
        Create an instance of the defensive distillation defence.
```

File: art/defences/transform/distillation.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+        :param classifier: A trained classifier.
+        :type classifier: :class:`.Classifier`
+        :param batch_size: Size of batches.
+        :type batch_size: `int`
+        :param nb_epochs: Number of epochs to use for training.
+        :type nb_epochs: `int`
+        """
+        super(Distillation, self).__init__(classifier=classifier)
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
        super().__init__(classifier=classifier)
```

File: run_tests.sh [hunk 16]
```diff
@@ -50,6 +50,7 @@ declare -a classifiers=("tests/classifiers/test_blackbox.py" \
 declare -a defences=("tests/defences/test_adversarial_trainer.py" \
                      "tests/defences/test_adversarial_trainer_madry_pgd.py" \
                      "tests/defences/test_class_labels.py" \
+                     "tests/defences/test_distillation.py" \
```
Comment (beat-buesser) [inline run_tests.sh]:
```suggestion
                     "tests/defences/test_defensive_distillation.py" \
```

File: art/defences/transform/distillation.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+        kwargs = {
+            "batch_size": batch_size,
+            "nb_epochs": nb_epochs,
+        }
+        self.set_params(**kwargs)
+
+    def __call__(self, x, modified_classifier):
+        """
+        Perform the transformation defence and return a robust classifier.
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
        Perform the defensive distillation defence mechanism and return a robuster classifier.
```

File: art/defences/transform/distillation.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+        :param x: Dataset for training the modified classifier.
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
        :param x: Dataset for training the transformed classifier.
```

File: art/defences/transform/distillation.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+        self._is_fitted = True
+
+        kwargs = {
+            "batch_size": batch_size,
+            "nb_epochs": nb_epochs,
+        }
+        self.set_params(**kwargs)
+
+    def __call__(self, x, modified_classifier):
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
    def __call__(self, x, transformed_classifier):
```

File: art/defences/transform/distillation.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+        """
+        Take in a dictionary of parameters and applies defence-specific checks before saving them as attributes.
+
+        :param batch_size: Size of batches.
+        :type batch_size: `int`
+        :param nb_epochs: Number of epochs to use for training.
+        :type nb_epochs: `int`
+        """
+        # Save defence-specific parameters
+        super(Distillation, self).set_params(**kwargs)
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
        super().set_params(**kwargs)
```

File: art/defences/transform/distillation.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+        are_probability = [is_probability(y) for y in modified_preds]
+        all_probability = np.sum(are_probability) == modified_preds.shape[0]
+
+        if not all_probability:
+            raise ValueError("The input modified classifier do not produce probability outputs.")
+
+        # Train the modified classifier with soft labels
+        modified_classifier.fit(x=x, y=preds, batch_size=self.batch_size, nb_epochs=self.nb_epochs)
+
+        return modified_classifier
```
Comment (beat-buesser) [inline art/defences/transform/distillation.py]:
```suggestion
        :param modified_classifier: A classifier to be transformed for increased robustness. Note that, the objective loss function
        used for fitting inside the input transformed_classifier must support soft labels, i.e. probability labels.
        :type modified_classifier: :class:`.Classifier`
        :return: The transformed classifier.
        :rtype: :class:`.Classifier`
        """
        # Check if the trained classifier produces probability outputs
        preds = self.classifier.predict(x=x, batch_size=self.batch_size)
        are_probability = [is_probability(y) for y in preds]
        all_probability = np.sum(are_probability) == preds.shape[0]

        if not all_probability:
            raise ValueError("The input trained classifier do not produce probability outputs.")

        # Check if the transformed classifier produces probability outputs
        transformed_preds = modified_classifier.predict(x=x, batch_size=self.batch_size)
        are_probability = [is_probability(y) for y in modified_preds]
        all_probability = np.sum(are_probability) == modified_preds.shape[0]

        if not all_probability:
            raise ValueError("The input modified classifier do not produce probability outputs.")

        # Train the transformed classifier with soft labels
        transformed_classifier.fit(x=x, y=preds, batch_size=self.batch_size, nb_epochs=self.nb_epochs)

        return transformed_classifier
```

File: art/defences/transform/transformer.py [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Get the internal classifier.
+
+        :return: The internal classifier.
+        :rtype: :class:`.Classifier`
+        """
+        return self.classifier
+
+    @abc.abstractmethod
+    def __call__(self, x, modified_classifier):
```
Comment (beat-buesser) [inline art/defences/transform/transformer.py]:
```suggestion
    def __call__(self, x, transformed_classifier):
```

File: art/defences/transform/transformer.py [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Perform the transformation defence and return a robust classifier.
+
+        :param x: Dataset for training the modified classifier.
+        :type x: `np.ndarray`
+        :param modified_classifier: A classifier to be modified for robustness.
+        :type modified_classifier: :class:`.Classifier`
+        :return: The modified classifier.
```
Comment (beat-buesser) [inline art/defences/transform/transformer.py]:
```suggestion
        Perform the transformation defence and return a robuster classifier.

        :param x: Dataset for training the transformed classifier.
        :type x: `np.ndarray`
        :param transformed_classifier: A classifier to be transformed for increased robustness.
        :type transformed_classifier: :class:`.Classifier`
        :return: The transformed classifier.
```

File: art/utils.py (hunk 6 above)
Comment (minhitbk) [reply art/utils.py]: Removed.

File: tests/test_utils.py (hunk 7 above)
Comment (minhitbk) [reply tests/test_utils.py]: Removed.

File: tests/test_utils.py (hunk 8 above)
Comment (minhitbk) [reply tests/test_utils.py]: Removed.

File: art/classifiers/tensorflow.py (hunk 10 above)
Comment (minhitbk) [reply art/classifiers/tensorflow.py]: Done.

File: art/classifiers/tensorflow.py (hunk 11 above)
Comment (minhitbk) [reply art/classifiers/tensorflow.py]: Done.

File: art/classifiers/tensorflow.py (hunk 12 above)
Comment (minhitbk) [reply art/classifiers/tensorflow.py]: Done.
